### PR TITLE
Minimal EIP2535 implementation of Inventory

### PR DESCRIPTION
**Pull request by zomglings**
This PR introduces a minimal EIP2535 implementation of an inventory into this repository. This implementation is adapted from the original implementation at https://github.com/G7DAO/contracts.

We have added this implementation here for two reasons:
1. Having an implementation in this repository makes our operations much simpler as we can access all our contracts using a single CLI. Previously, we were interacting with the inventory with the `game7ctl` CLI while all other contract interactions were done with the CLI in this repo.
2. We have made several simplifications to the Inventory contract that might break Game7's deployment pipelines. This is less painful for all involved.

Resolves https://github.com/moonstream-to/web3/issues/319 and https://github.com/moonstream-to/web3/issues/322

### CODE REVIEW

File: cli/web3cli/test_inventory.py [hunk 1]
```diff
@@ -0,0 +1,2165 @@
...
+            item_unequipped_events[0]["args"]["amount"],
+            2,
+        )
+        self.assertEqual(
+            item_unequipped_events[0]["args"]["unequippedBy"],
+            self.player.address,
+        )
+
+    def test_player_cannot_unequip_erc20_tokens_from_persistent_slot_but_can_increase_amount(self):
+        """
```
Comment (kellan-simiotics) [inline cli/web3cli/test_inventory.py:2075]: I think we should change the test name to indicate the test is broken or just comment out the whole test for now. I read through equip / _unequip methods a couple times before I noticed the comment that noted this test was unfinished as the functionality was missing.

**Review by kellan-simiotics — CHANGES_REQUESTED** [review]
I'm not sure if there is anything critical missing, but the contract is rather difficult to use if you can ask a slot for all of its equippable options. Ran into this difficulty during CG testing.

File: contracts/inventory/IInventory.sol [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+// Calculated by solface: https://github.com/moonstream-to/solface
+//
+// To recalculate from root directory of this repo:
+// $ jq .abi build/contracts/IInventory.json  | solface -name IInventory -annotations | grep "Interface ID:"
+//
+// Note: Change path to build/contracts/IInventory.json depending on where you are relative to the repo root.
+interface IInventory {
+    // This event should be emitted when the subject ERC721 contract address is set (or changes) on the
+    // Inventory contract.
+    event NewSubjectAddress(address indexed contractAddress);
```
[the remark this answers is not in the document]
Comment (zomglings) [reply contracts/inventory/IInventory.sol:28]: It depends on the implementation of the interface. All we are specifying in the interface definition is that *if you change the subject address* you should emit this event.

File: contracts/inventory/InventoryFacet.sol [hunk 3]
```diff
@@ -282,7 +234,14 @@ contract InventoryFacet is
         address itemAddress,
         uint256 itemPoolId,
         uint256 maxAmount
-    ) external onlyAdmin requireValidItemType(itemType) {
+    ) external onlyAdmin {
```
[the remark this answers is not in the document]
Comment (zomglings) [reply contracts/inventory/InventoryFacet.sol:237]: I don't even know how to expose this on-chain without adding an enumeration.

That can be an extension of the base interface (see `IERC721Enumerable`, which is an extension of `IERC721`). I personally prefer to get that information from events. I don't think it's necessary to the system.

File: contracts/inventory/InventoryFacet.sol [hunk 4]
```diff
@@ -246,34 +196,36 @@ contract InventoryFacet is
...
 
-        LibInventory.Slot memory slot = istore.SlotData[slotId];
+        Slot memory slot = istore.SlotData[slotId];
         slot.SlotURI = newSlotURI;
         istore.SlotData[slotId] = slot;
         emit NewSlotURI(slotId);
     }
 
-    function slotIsUnequippable(uint256 slotId) external view returns (bool) {
+    function slotIsPersistent(uint256 slotId) external view returns (bool) {
```
[the remark this answers is not in the document]
Comment (zomglings) [reply contracts/inventory/InventoryFacet.sol:212]: This name was suggested by Diniz from OP Games. He subclassed `InventoryFacet` for their usage.

It is also a name that people have responded well to.

I suggest we merge this PR in with `Persistent` as the name, and we can change it later based on research and feedback.

[the remark this answers is not in the document]
Comment (kellan-simiotics) [reply contracts/inventory/InventoryFacet.sol:212]: +1

File: contracts/inventory/IInventory.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (kellan-simiotics) [reply contracts/inventory/IInventory.sol:28]: It seems like this ventures into the realm of speculating about functionality that isn't really in the spirit of the existing code.

File: cli/web3cli/test_inventory.py [hunk 5]
```diff
@@ -0,0 +1,2165 @@
...
+        )
+        slot = self.inventory.num_slots()
+
+        erc721_type = 721
+
+        with self.assertRaises(VirtualMachineError):
+            self.inventory.mark_item_as_equippable_in_slot(
+                slot,
+                erc721_type,
+                self.payment_token.address,
```
[the remark this answers is not in the document]
Comment (zomglings) [reply cli/web3cli/test_inventory.py]: Nice catch. Fixed.

File: cli/web3cli/test_inventory.py [hunk 6]
```diff
@@ -0,0 +1,2165 @@
...
```
[the remark this answers is not in the document]
Comment (zomglings) [reply cli/web3cli/test_inventory.py]: Fixed.

File: contracts/inventory/InventoryFacet.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (kellan-simiotics) [reply contracts/inventory/InventoryFacet.sol:237]: Yes, but contract administrators won't always have easy access to the events. Can't we just dump the SlotEligibleItems object?

File: cli/web3cli/test_inventory.py (hunk 1 above)
Comment (zomglings) [reply cli/web3cli/test_inventory.py:2075]: I renamed the test to `test_player_cannot_unequip_erc20_tokens_from_persistent_slot`.

Comment (zomglings) [reply cli/web3cli/test_inventory.py:2075]: It's still a useful test as it tests persistence of slots.

**Review by kellan-simiotics — APPROVED** [review]
lg

**Pull request merged**